storage2/src/state.rs

```diff
@@ -0,0 +1,47 @@
+use std::collections::HashMap;
+
+use jmt::{
+    storage::{TreeReader, TreeWriter},
+    JellyfishMerkleTree,
+};
+
+mod transaction;
+use transaction::Transaction as StateTransaction;
+
+/// State is a lightweight copy-on-write fork of the chain state,
+/// implemented as a RYW cache over a pinned JMT version.
+pub(crate) struct State<'a, R: TreeReader + TreeWriter> {
+    cache: HashMap<jmt::KeyHash, jmt::OwnedValue>,
+    jmt_version: jmt::Version,
+    jmt: &'a JellyfishMerkleTree<'a, R>,
+}
+
+impl<'a, R: TreeReader + TreeWriter> State<'a, R> {
+    pub fn new() -> Self {
+        Self {
+            cache: todo!(),
+            jmt_version: todo!(),
+            jmt: todo!(),
+        }
+    }
+
+    pub fn begin_transaction(&mut self) -> StateTransaction<'a, R> {
+        StateTransaction::new()
+    }
+}
+
+impl<'a, R: TreeReader + TreeWriter> StateRead for State<'a, R> {
+    fn get(&self, key: jmt::KeyHash) -> Option<&jmt::OwnedValue> {
+        todo!()
+    }
+}
+
+pub trait StateRead {
+    /// Get
+    fn get(&self, key: jmt::KeyHash) -> Option<&jmt::OwnedValue>;
+}
```

Comment:
Per #1454, a goal of the redesign is that we want to be operating on string keys, so we shouldn't be taking a `jmt::KeyHash` here, we should be taking a string key. Since we don't need to hold on to the key past the length of the query, we don't need to take ownership, so that could be a `key: &str`.

We don't want to return a `jmt::OwnedValue`, because we don't want API users to have to reference the JMT API. We could return a `Vec<u8>` instead.

Reply:
Also, do we want the get method to return a byte vector at all? That's in some sense the "underlying" operation, but it's also the least frequently used. We mostly want domain types, or sometimes proto types.

What if instead, we renamed the byte-returning method as get_raw (a required method), and then had provided methods get_proto (returning a proto type) and get (returning a domain type).

Then the trait implementor would only have to implement get_raw, but the user could call get and work with domain types.

Reply:
Finally, I think that this method can't be a synchronous read, but will have to be an async one. We could either use the async_trait macro, or return an associated type. Using async_trait seems easier to handle the provided methods.